core/common/src/main/java/alluxio/util/executor/SerializedTaskRunner.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.util.executor;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class can execute a series of tasks on a separate thread in the order they are submitted.
+ * No guarantees can be made about the timing of how the tasks complete, or what the result of
+ * the tasks are.
+ *
+ * All tasks in the queue are guaranteed to complete before {@link #shutdown(long, TimeUnit)} is
+ * called so long as they complete within the provided timeout. At which time the task runner will
+ * stop accepting new tasks (all calls to {@link #addTask(Runnable)} will return false).
+ *
+ */
+public class SerializedTaskRunner implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(SerializedTaskRunner.class);
+
+  private volatile boolean mRunning;
+  private volatile boolean mIsShutdown;
+  private final LinkedBlockingQueue<Runnable> mTasks;
+  private volatile Thread mRunningThread;
+
+  /**
+   * Create a new instance of {@link SerializedTaskRunner}.
+   */
+  public SerializedTaskRunner() {
+    mTasks = new LinkedBlockingQueue<>();
+    mRunning = true;
+    mIsShutdown = false;
+  }
+
+  /**
+   * Add a new task to execute into the queue.
+   *
+   * @param r task to run
+   *
+   * @return true if the task was added to the queue successfully
+   */
+  public boolean addTask(Runnable r) {
+    if (mIsShutdown) {
+      return false;
+    }
+    return mTasks.add(r);
+  }
+
+  @Override
+  public void run() {
+    mRunningThread = Thread.currentThread();
+    while (mRunning) {
+      try {
+        Runnable task = mTasks.take();
+        try {
+          task.run();
+        } catch (RuntimeException e) {
```

Comment:
Throwable

Reply:
Updated.

Also, I apologize if this conflicts with your other PR. Correct me if I'm wrong, I don't think we're solving the same problems. We recently found the issue which I'm trying to solve in this PR. I guess the timing is just unfortunate.